subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <errno.h>
+#include <init.h>
+#include <stdint.h>
+#include <zephyr.h>
+#include <net/buf.h>
+#include <logging/log.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/services/ias.h>
+
+#define LOG_LEVEL CONFIG_BT_IAS_CLIENT_LOG_LEVEL
+LOG_MODULE_REGISTER(ias_client);
+
+#define IS_HANDLE_VALID(handle) ((handle) != 0x0000)
+
+static struct bt_ias_client_cb *ias_client_cb;
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+int bt_ias_alert_write(struct bt_conn *conn, uint8_t alert_lvl)
+{
+	int err;
+	struct net_buf_simple buf;
+
+	net_buf_simple_init_with_data(&buf, &alert_lvl, 1);
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write.handle,
+					buf.data, buf.len, false);
+	if (err < 0) {
+		BT_ERR("IAS client alert level write failed");
+		return err;
+	}
+
+	return 0;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	client_by_conn(conn)->write.handle = attr->handle;
+	ias_client_cb->discover();
+	return 0;
+}
+
+static uint8_t bt_ias_prim_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	int err = 0;
+	struct bt_gatt_service_val *data;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (!attr) {
+		BT_DBG("Discover complete\n");
+		return BT_GATT_ITER_STOP;
+	}
+	data = attr->user_data;
+
+	(void)memcpy(&client->uuid, BT_UUID_ALERT_LEVEL, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.start_handle = sys_cpu_to_le16(attr->handle + 1);
+	client->discover.end_handle = sys_cpu_to_le16(data->end_handle);
+	client->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	client->discover.func = bt_ias_alert_lvl_disc_cb;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err) {
+		BT_ERR("Characteristic discover failed\n");
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_ias_discover(struct bt_conn *conn)
+{
+	int err;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	(void)memcpy(&client->uuid, BT_UUID_IAS, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.func = bt_ias_prim_disc_cb;
+	client->discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	client->discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	client->discover.type = BT_GATT_DISCOVER_PRIMARY;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err < 0) {
+		BT_ERR("Primary discover failed(err %d\n)", err);
+		return err;
+	}
+
+	return BT_GATT_ITER_STOP;
```

Comment:
The return value is invalid here I guess.

Reply:
Done